File: computing_methods/SystemOfLinearEquations/src/SeidelMethod.java
import javafx.util.Pair;

public class SeidelMethod implements SolvingMethod {
    public final int iterationsLimit;

    public SeidelMethod(int iterationsLimit) {
        this.iterationsLimit = iterationsLimit;
    }

    @Override
    public Pair<Vector, Integer> solve(LinearEquationsSystem system, double epsilon) {
        Matrix A = system.getA();
        Vector b = system.getB();
        int n = b.getDimensions();
        Vector x = Vector.zero(n);
        Vector nextX = x;
        int iterations = 0;
        do {
            x = nextX;
            double[] temp = new double[A.getHeight()];
            for (int i = 0; i < n; i++) {
                double sum = 0.;
                for (int j = 0; j < i; j++) {
                    sum += (-A.get(i, j)) * temp[j];
                }
                for (int j = i + 1; j < n; j++) {
                    sum += (-A.get(i, j)) * x.get(j);
                }
                sum += b.get(i);
                sum /= A.get(i, i);
                temp[i] = sum;
            }
            nextX = new Vector(temp);
            iterations++;
            if (nextX.hasNaN() || iterations > iterationsLimit)
                return null;
        } while (A.multiply(x).subtract(b).norm() > epsilon * 0.1);
        return !(A.multiply(x).subtract(b).norm() <= epsilon * 0.1) && !(Vector.distance(x, nextX) <= epsilon * 0.1) ? null : new Pair<>(nextX, iterations);
    }
}

File: computing_methods/CourseWork/src/SeidelMethod.scala
class SeidelMethod(iterationsLimit: Int) extends LinearSystemsSolvingMethod {
  override def solve(s: LinearEquationSystem, epsilon: Double): (Vector, Int) = {
    val A = s.getA
    val b = s.getB
    val n = b.getDimension
    var x = Vector.zero(n)
    var nextX = x
    var iterations = 0
    do {
      x = nextX
      val temp = Array.ofDim[Double](A.getHeight)
      for (i <- 0 until n) {
        var sum = 0d
        for (j <- 0 until i) sum = sum + (-A(i, j)) * temp(j)
        for (j <- i + 1 until n) sum = sum + (-A(i, j)) * x(j)
        temp update(i, (sum + b(i)) / A(i, i))
      }
      nextX = new Vector(temp)
      iterations = iterations + 1
      if (nextX.hasNan || iterations > iterationsLimit) return null
    } while ((A * x - b).norm > epsilon * 0.1)
    if (!(A * x - b).norm.<=(epsilon * 0.1) && !Vector.distance(x, nextX).<=(epsilon * 0.1)) null
    else (nextX, iterations)
  }
}
